### ISSUE TITLE
unused parameters when using WavLM. cased crash when using DDP

### ISSUE THREAD

**Issue by avishaiElmakies** [speechbrain/speechbrain#2340]
### Describe the bug

when trying to finetune WavLM and using DDP. there are some unused parameters. This causes the run to crash. when using --find_unused_parameters it says there are no unused parameters. but when giving detail it says names of parameters that are not used. and when checked using code after forward it indeed show the grad is none after forward.

### Expected behaviour

all parameters should be used

### To Reproduce

yaml 

```
########
# used to finetune a wavlm-large model on emotion from scratch
# created by avishai elmakies 15.1.2024
########



# Seed needs to be set at top of yaml, before objects with parameters are made
seed: 12
__set_seed: !apply:torch.manual_seed [12]
output_folder: finetune_e_wavlm_large
# eder_file: finetune_emo/eder.txt
save_folder: &id008 !ref <output_folder>/save
train_log: &id009 !ref <output_folder>/train_log.txt
local-rank: 0
distributed_launch: false

## some important costants 
sample_rate: 16000
download_base_path: "../../models"
dataset_json: ../../data_jsons/emotion_finetune_dataset.json
split_ratio: [0.8, 0.1, 0.1]

window_length: 1 # win_len = 0.02 * 1 = 0.02s
stride: 1 # stride = 0.02 * 1 = 0.02s

encoder_dim: 1024
# Outputs
out_n_neurons: 4  # BPE size, index(blank/eos/bos) = 0

# Dataloader options
# With data_parallel batch_size is split into N jobs
# With DDP batch_size is multiplied by N jobs
dataloader_options:
  batch_size: 2
  shuffle: true
  num_workers: 2    # 2 on linux but 0 works on windows
  drop_last: false
  pin_memory: true
  collate_fn: !name:speechbrain.dataio.batch.PaddedBatch

test_dataloader_opts:
  batch_size: 2
  collate_fn: !name:speechbrain.dataio.batch.PaddedBatch

epoch_counter: &id007 !new:speechbrain.utils.epoch_loop.EpochCounter

  limit: 1


add_noise_aug: &id020 !new:speechbrain.processing.speech_augmentation.AddNoise
  snr_low: 10
  snr_high: 20
  mix_prob: 0.5 


drop_freq_aug: &id021 !new:speechbrain.processing.speech_augmentation.DropFreq
  drop_freq_high: 0.5
  drop_count_low: 2
  drop_count_high: 4
  drop_prob : 0.5



drop_chunk_aug: &id022 !new:speechbrain.processing.speech_augmentation.DropChunk
  drop_length_low : 1600
  drop_length_high : 16000
  drop_count_low: 1
  drop_count_high: 2


augmentation: !new:speechbrain.processing.augmentation.Augmenter
  parallel_augment: false
  concat_original: false
  min_augmentations: 0
  max_augmentations: 3
  repeat_augment: 1
  noise_aug: *id020
  freq_aug: *id021
  chunk_aug: *id022

input_norm: &id001 !new:speechbrain.processing.features.InputNormalization
  norm_type: sentence
  std_norm: false


wav2vec2: &id002 !new:speechbrain.lobes.models.huggingface_wav2vec.HuggingFaceWav2Vec2
  source: microsoft/wavlm-large
  output_norm: true
  freeze: false
  freeze_feature_extractor: true
  save_path: !ref <download_base_path>/wavlm-large

avg_pool: !new:speechbrain.nnet.pooling.Pooling1d
  pool_type: avg
  kernel_size: 1
  stride: 1
  ceil_mode: true

output_mlp: &id003 !new:speechbrain.nnet.linear.Linear
  input_size: !ref <encoder_dim>
  n_neurons: !ref <out_n_neurons>
  bias: false

log_softmax: !new:speechbrain.nnet.activations.Softmax
  apply_log: true

compute_cost: !name:emotion.finetune.utils.weighted_nll_loss

# can be used with compute loss, probably better for ddp to work like this
# should also know the labels in advance. can work with diffrent labels and label encoeders
weights : 
  "s": 8.0
  "h": 20.0
  "n": 1.0
  "a": 20.0

modules:
  input_norm: *id001
  wav2vec2: *id002
  output_mlp: *id003

opt_class: !name:torch.optim.Adam
  lr: 0.0001

wav2vec2_opt_class: !name:torch.optim.Adam
  lr: 0.00001

lr_annealing: &id005 !new:speechbrain.nnet.schedulers.NewBobScheduler
  initial_value: 0.0001
  improvement_threshold: 0.0025
  annealing_factor: 0.8
  patient: 0

lr_annealing_wav2vec2: &id006 !new:speechbrain.nnet.schedulers.NewBobScheduler
  initial_value: 0.00001
  improvement_threshold: 0.0025
  annealing_factor: 0.9
  patient: 0


checkpointer: !new:speechbrain.utils.checkpoints.Checkpointer
  checkpoints_dir: *id008
  recoverables:
    scheduler_model: *id005
    scheduler_wav2vec: *id006
    counter: *id007
train_logger: !new:speechbrain.utils.train_logger.FileTrainLogger
  save_file: *id009

error_stats: !name:speechbrain.utils.metric_stats.ClassificationStats
```

code:
[Code.zip](https://github.com/speechbrain/speechbrain/files/13953849/Code.zip)

### Environment Details

```
SpeechBrain system description
==============================
Python version:
3.9.2 (default, Feb 28 2021, 17:03:44) 
[GCC 10.2.1 20210110]
==============================
Installed Python packages:
aiohttp==3.9.1
aiosignal==1.3.1
alembic==1.13.0
AMFM-decompy==1.0.11
annotated-types==0.6.0
antlr4-python3-runtime==4.8
asteroid-filterbanks==0.4.0
asttokens==2.4.1
async-timeout==4.0.3
attrs==23.1.0
audioread==3.0.1
bitarray==2.9.0
blis==0.7.11
catalogue==2.0.10
certifi==2022.12.7
cffi==1.16.0
charset-normalizer==2.1.1
click==8.1.7
cloudpathlib==0.16.0
colorama==0.4.6
coloredlogs==15.0.1
colorlog==6.8.0
comm==0.2.0
confection==0.1.4
contourpy==1.2.0
cycler==0.12.1
cymem==2.0.8
Cython==3.0.6
datasets==2.15.0
debugpy==1.8.0
decorator==5.1.1
dill==0.3.7
docopt==0.6.2
einops==0.7.0
en-core-web-sm @ https://github.com/explosion/spacy-models/releases/download/en_core_web_sm-3.7.1/en_core_web_sm-3.7.1-py3-none-any.whl#sha256=86cc141f63942d4b2c5fcee06630fd6f904788d2f0ab005cce45aadb8fb73889
exceptiongroup==1.2.0
executing==2.0.1
fairseq @ git+https://github.com/pytorch/fairseq@da8fb630880d529ab47e53381c30ddc8ad235216
filelock==3.9.0
flatbuffers==23.5.26
fonttools==4.46.0
frozenlist==1.4.0
fsspec==2023.10.0
greenlet==3.0.2
huggingface-hub==0.19.4
humanfriendly==10.0
hydra-core==1.0.7
HyperPyYAML==1.2.2
idna==3.4
importlib-metadata==7.0.0
importlib-resources==6.1.1
inflect==7.0.0
iniconfig==2.0.0
ipykernel==6.27.1
ipython==8.18.1
jedi==0.19.1
Jinja2==3.1.2
joblib==1.3.2
julius==0.2.7
jupyter_client==8.6.0
jupyter_core==5.5.1
kiwisolver==1.4.5
langcodes==3.3.0
lazy_loader==0.3
librosa==0.10.0
lightning==2.1.2
lightning-utilities==0.10.0
llvmlite==0.36.0
lxml==4.9.3
Mako==1.3.0
markdown-it-py==3.0.0
MarkupSafe==2.1.3
matplotlib==3.8.2
matplotlib-inline==0.1.6
mdurl==0.1.2
mpmath==1.3.0
msgpack==1.0.7
multidict==6.0.4
multiprocess==0.70.15
murmurhash==1.0.10
nest-asyncio==1.5.8
networkx==3.0
npy-append-array==0.9.16
numba==0.53.0
numpy==1.22.0
omegaconf==2.0.6
onnxruntime==1.16.3
optuna==3.5.0
packaging==23.2
pandas==2.1.4
parso==0.8.3
pexpect==4.9.0
Pillow==9.3.0
platformdirs==4.0.0
pluggy==1.3.0
pooch==1.8.0
portalocker==2.8.2
preshed==3.0.9
primePy==1.3
prompt-toolkit==3.0.43
protobuf==4.25.1
psutil==5.9.7
ptyprocess==0.7.0
pure-eval==0.2.2
pyannote.audio==3.1.1
pyannote.core==5.0.0
pyannote.database==5.0.1
pyannote.metrics==3.2.1
pyannote.pipeline==3.0.1
pyarrow==14.0.1
pyarrow-hotfix==0.6
pycparser==2.21
pydantic==2.5.2
pydantic_core==2.14.5
Pygments==2.17.2
pyparsing==3.1.1
pytest==7.4.3
python-dateutil==2.8.2
pytorch-lightning==2.1.2
pytorch-metric-learning==2.3.0
pytz==2023.3.post1
PyYAML==6.0.1
pyzmq==25.1.2
regex==2023.10.3
requests==2.28.1
resampy==0.4.2
rich==13.7.0
ruamel.yaml==0.18.5
ruamel.yaml.clib==0.2.8
sacrebleu==2.4.0
safetensors==0.4.1
scikit-learn==1.3.2
scipy==1.11.4
semver==3.0.2
sentencepiece==0.1.99
shellingham==1.5.4
simplejson==3.19.2
six==1.16.0
smart-open==6.4.0
sortedcontainers==2.4.0
soundfile==0.12.1
soxr==0.3.7
spacy==3.7.2
spacy-legacy==3.0.12
spacy-loggers==1.0.5
speechbrain==0.5.16
SQLAlchemy==2.0.23
srsly==2.4.8
stack-data==0.6.3
sympy==1.12
tabulate==0.9.0
tensorboardX==2.6.2.2
-e git+https://github.com/facebookresearch/textlesslib.git@ba33d669d8284b4f7bfe81e7384e83ab799fe384#egg=textless
tgt==1.5
thinc==8.2.1
threadpoolctl==3.0.0
tokenizers==0.15.0
tomli==2.0.1
torch==2.1.1+cu118
torch-audiomentations==0.11.0
torch-pitch-shift==1.2.4
torchaudio==2.1.1+cu118
torchcrepe==0.0.22
torchmetrics==1.2.1
torchvision==0.16.1+cu118
tornado==6.4
tqdm==4.66.1
traitlets==5.14.0
transformers==4.36.1
triton==2.1.0
typer==0.9.0
typing_extensions==4.9.0
tzdata==2023.3
Unidecode==1.3.7
urllib3==1.26.13
wasabi==1.1.2
wcwidth==0.2.12
weasel==0.3.4
wget==3.2
xxhash==3.4.1
yarl==1.9.4
zipp==3.17.0
==============================
Could not get git revision==============================
CUDA version:
11.8
```

### Relevant Log Output

```shell
Traceback:

Traceback (most recent call last):
  File "/cs/labs/oabend/avishai.elma/src/emotion_src/emotion_finetune.py", line 63, in <module>
    main()
  File "/cs/labs/oabend/avishai.elma/src/emotion_src/emotion_finetune.py", line 46, in main
    brain = finetune(modules, hparams, datasets,class_weights=class_weights, run_opts=run_opts,
  File "/cs/labs/oabend/avishai.elma/src/emotion_src/emotion/finetune/utils.py", line 136, in finetune
    emo_brain.fit(
  File "/cs/labs/oabend/avishai.elma/lab_env_v2/lib/python3.9/site-packages/speechbrain/core.py", line 1366, in fit
    self._fit_train(train_set=train_set, epoch=epoch, enable=enable)
  File "/cs/labs/oabend/avishai.elma/lab_env_v2/lib/python3.9/site-packages/speechbrain/core.py", line 1193, in _fit_train
    loss = self.fit_batch(batch)
  File "/cs/labs/oabend/avishai.elma/src/emotion_src/emotion/finetune/finetune_brain.py", line 58, in fit_batch
    predictions = self.compute_forward(batch, sb.Stage.TRAIN)
  File "/cs/labs/oabend/avishai.elma/src/emotion_src/emotion/finetune/finetune_brain.py", line 33, in compute_forward
    outputs = self.modules.wav2vec2(wavs)
  File "/cs/labs/oabend/avishai.elma/lab_env_v2/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/cs/labs/oabend/avishai.elma/lab_env_v2/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/cs/labs/oabend/avishai.elma/lab_env_v2/lib/python3.9/site-packages/torch/nn/parallel/distributed.py", line 1515, in forward
    inputs, kwargs = self._pre_forward(*inputs, **kwargs)
  File "/cs/labs/oabend/avishai.elma/lab_env_v2/lib/python3.9/site-packages/torch/nn/parallel/distributed.py", line 1409, in _pre_forward
    if torch.is_grad_enabled() and self.reducer._rebuild_buckets():
RuntimeError: Expected to have finished reduction in the prior iteration before starting a new one. This error indicates that your module has parameters that were not used in producing loss. You can enable unused parameter detection by passing the keyword argument `find_unused_parameters=True` to `torch.nn.parallel.DistributedDataParallel`, and by 
making sure all `forward` function outputs participate in calculating loss. 
If you already have done the above, then the distributed data parallel module wasn't able to locate the output tensors in the return value of your module's `forward` function. Please include the loss function and the structure of the return value of `forward` of your module when reporting this issue (e.g. list, dict, iterable).
Parameters which did not receive grad for rank 0: model.encoder.layers.22.final_layer_norm.bias, model.encoder.layers.22.final_layer_norm.weight, model.encoder.layers.22.feed_forward.output_dense.bias, model.encoder.layers.22.feed_forward.output_dense.weight, model.encoder.layers.22.feed_forward.intermediate_dense.bias, model.encoder.layers.22.feed_forward.intermediate_dense.weight, model.encoder.layers.22.layer_norm.bias, model.encoder.layers.22.layer_norm.weight, model.encoder.layers.22.attention.gru_rel_pos_linear.bias, model.encoder.layers.22.attention.gru_rel_pos_linear.weight, model.encoder.layers.22.attention.out_proj.bias, model.encoder.layers.22.attention.out_proj.weight, model.encoder.layers.22.attention.q_proj.bias, model.encoder.layers.22.attention.q_proj.weight, model.encoder.layers.22.attention.v_proj.bias, model.encoder.layers.22.attention.v_proj.weight, model.encoder.layers.22.attention.k_proj.bias, model.encoder.layers.22.attention.k_proj.weight, model.encoder.layers.22.attention.gru_rel_pos_const, model.masked_spec_embed
Parameter indices which did not receive grad for rank 0: 0 429 430 431 432 433 434 435 436 437 438 439 440 441 442 443 444 445 446 447
Traceback (most recent call last):
  File "/cs/labs/oabend/avishai.elma/src/emotion_src/emotion_finetune.py", line 63, in <module>
    main()
  File "/cs/labs/oabend/avishai.elma/src/emotion_src/emotion_finetune.py", line 46, in main
    brain = finetune(modules, hparams, datasets,class_weights=class_weights, run_opts=run_opts,
  File "/cs/labs/oabend/avishai.elma/src/emotion_src/emotion/finetune/utils.py", line 136, in finetune
    emo_brain.fit(
  File "/cs/labs/oabend/avishai.elma/lab_env_v2/lib/python3.9/site-packages/speechbrain/core.py", line 1366, in fit
    self._fit_train(train_set=train_set, epoch=epoch, enable=enable)
  File "/cs/labs/oabend/avishai.elma/lab_env_v2/lib/python3.9/site-packages/speechbrain/core.py", line 1193, in _fit_train
    loss = self.fit_batch(batch)
  File "/cs/labs/oabend/avishai.elma/src/emotion_src/emotion/finetune/finetune_brain.py", line 58, in fit_batch
    predictions = self.compute_forward(batch, sb.Stage.TRAIN)
  File "/cs/labs/oabend/avishai.elma/src/emotion_src/emotion/finetune/finetune_brain.py", line 33, in compute_forward
    outputs = self.modules.wav2vec2(wavs)
  File "/cs/labs/oabend/avishai.elma/lab_env_v2/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/cs/labs/oabend/avishai.elma/lab_env_v2/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/cs/labs/oabend/avishai.elma/lab_env_v2/lib/python3.9/site-packages/torch/nn/parallel/distributed.py", line 1515, in forward
    inputs, kwargs = self._pre_forward(*inputs, **kwargs)
  File "/cs/labs/oabend/avishai.elma/lab_env_v2/lib/python3.9/site-packages/torch/nn/parallel/distributed.py", line 1409, in _pre_forward
    if torch.is_grad_enabled() and self.reducer._rebuild_buckets():
RuntimeError: Expected to have finished reduction in the prior iteration before starting a new one. This error indicates that your module has parameters that were not used in producing loss. You can enable unused parameter detection by passing the keyword argument `find_unused_parameters=True` to `torch.nn.parallel.DistributedDataParallel`, and by 
making sure all `forward` function outputs participate in calculating loss. 
If you already have done the above, then the distributed data parallel module wasn't able to locate the output tensors in the return value of your module's `forward` function. Please include the loss function and the structure of the return value of `forward` of your module when reporting this issue (e.g. list, dict, iterable).
Parameters which did not receive grad for rank 1: model.encoder.layers.22.final_layer_norm.bias, model.encoder.layers.22.final_layer_norm.weight, model.encoder.layers.22.feed_forward.output_dense.bias, model.encoder.layers.22.feed_forward.output_dense.weight, model.encoder.layers.22.feed_forward.intermediate_dense.bias, model.encoder.layers.22.feed_forward.intermediate_dense.weight, model.encoder.layers.22.layer_norm.bias, model.encoder.layers.22.layer_norm.weight, model.encoder.layers.22.attention.gru_rel_pos_linear.bias, model.encoder.layers.22.attention.gru_rel_pos_linear.weight, model.encoder.layers.22.attention.out_proj.bias, model.encoder.layers.22.attention.out_proj.weight, model.encoder.layers.22.attention.q_proj.bias, model.encoder.layers.22.attention.q_proj.weight, model.encoder.layers.22.attention.v_proj.bias, model.encoder.layers.22.attention.v_proj.weight, model.encoder.layers.22.attention.k_proj.bias, model.encoder.layers.22.attention.k_proj.weight, model.encoder.layers.22.attention.gru_rel_pos_const, model.masked_spec_embed
Parameter indices which did not receive grad for rank 1: 0 429 430 431 432 433 434 435 436 437 438 439 440 441 442 443 444 445 446 447
[2024-01-17 21:13:33,097] torch.distributed.elastic.multiprocessing.api: [ERROR] failed (exitcode: 1) local_rank: 0 (pid: 2021552) of binary: /cs/labs/oabend/avishai.elma/lab_env_v2/bin/python
Traceback (most recent call last):
  File "/usr/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/cs/labs/oabend/avishai.elma/lab_env_v2/lib/python3.9/site-packages/torch/distributed/launch.py", line 196, in <module>
    main()
  File "/cs/labs/oabend/avishai.elma/lab_env_v2/lib/python3.9/site-packages/torch/distributed/launch.py", line 192, in main
    launch(args)
  File "/cs/labs/oabend/avishai.elma/lab_env_v2/lib/python3.9/site-packages/torch/distributed/launch.py", line 177, in launch
    run(args)
  File "/cs/labs/oabend/avishai.elma/lab_env_v2/lib/python3.9/site-packages/torch/distributed/run.py", line 797, in run
    elastic_launch(
  File "/cs/labs/oabend/avishai.elma/lab_env_v2/lib/python3.9/site-packages/torch/distributed/launcher/api.py", line 134, in __call__
    return launch_agent(self._config, self._entrypoint, list(args))
  File "/cs/labs/oabend/avishai.elma/lab_env_v2/lib/python3.9/site-packages/torch/distributed/launcher/api.py", line 264, in launch_agent
    raise ChildFailedError(
torch.distributed.elastic.multiprocessing.errors.ChildFailedError: 
============================================================
emotion_finetune.py FAILED
------------------------------------------------------------
Failures:
[1]:
  time      : 2024-01-17_21:13:33
  host      : firth-02.cs.huji.ac.il
  rank      : 1 (local_rank: 1)
  exitcode  : 1 (pid: 2021553)
  error_file: <N/A>
  traceback : To enable traceback see: https://pytorch.org/docs/stable/elastic/errors.html
------------------------------------------------------------
Root Cause (first observed failure):
[0]:
  time      : 2024-01-17_21:13:33
  host      : firth-02.cs.huji.ac.il
  rank      : 0 (local_rank: 0)
  exitcode  : 1 (pid: 2021552)
  error_file: <N/A>
  traceback : To enable traceback see: https://pytorch.org/docs/stable/elastic/errors.html
============================================================
```

**Comment by BenoitWang** [comment]
Hi @avishaiElmakies, I didn't use ddp for training and I didn't see this before. I did some research and I found: an [explanation](https://github.com/pytorch/pytorch/issues/43259#issuecomment-1038180530) for this error, and some possible causes and solutions [here](https://github.com/pytorch/pytorch/issues/43259). I guess possibly there's an incorrect `torch.no_grad()` in your code. However I suggest first launch ddp with the [original recipe](https://github.com/speechbrain/speechbrain/tree/develop/recipes/ZaionEmotionDataset), if it works then compare with your  own modifications to debug.

**Comment by avishaiElmakies** [comment]
hi @BenoitWang i just ran your original recipe and i am getting the same thing. (I took the new develop branch as well)

Traceback:

```
[2024-01-18 13:11:13,288] torch.distributed.run: [WARNING] 
[2024-01-18 13:11:13,288] torch.distributed.run: [WARNING] *****************************************
[2024-01-18 13:11:13,288] torch.distributed.run: [WARNING] Setting OMP_NUM_THREADS environment variable for each process to be 1 in default, to avoid your system being overloaded, please further tune the variable for optimal performance in your application as needed. 
[2024-01-18 13:11:13,288] torch.distributed.run: [WARNING] *****************************************
[W socket.cpp:436] [c10d] The server socket cannot be initialized on [::]:29500 (errno: 97 - Address family not supported by protocol).
[W socket.cpp:663] [c10d] The client socket cannot be initialized to connect to [::ffff:127.0.0.1]:29500 (errno: 97 - Address family not supported by protocol).
[W socket.cpp:663] [c10d] The client socket cannot be initialized to connect to [::ffff:127.0.0.1]:29500 (errno: 97 - Address family not supported by protocol).
[W socket.cpp:663] [c10d] The client socket cannot be initialized to connect to [::ffff:127.0.0.1]:29500 (errno: 97 - Address family not supported by protocol).
Some weights of the model checkpoint at microsoft/wavlm-large were not used when initializing WavLMModel: ['encoder.pos_conv_embed.conv.weight_v', 'encoder.pos_conv_embed.conv.weight_g']
- This IS expected if you are initializing WavLMModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing WavLMModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of WavLMModel were not initialized from the model checkpoint at microsoft/wavlm-large and are newly initialized: ['encoder.pos_conv_embed.conv.parametrizations.weight.original1', 'encoder.pos_conv_embed.conv.parametrizations.weight.original0']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
Some weights of the model checkpoint at microsoft/wavlm-large were not used when initializing WavLMModel: ['encoder.pos_conv_embed.conv.weight_v', 'encoder.pos_conv_embed.conv.weight_g']
- This IS expected if you are initializing WavLMModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing WavLMModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of WavLMModel were not initialized from the model checkpoint at microsoft/wavlm-large and are newly initialized: ['encoder.pos_conv_embed.conv.parametrizations.weight.original1', 'encoder.pos_conv_embed.conv.parametrizations.weight.original0']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
speechbrain.lobes.models.huggingface_transformers.wav2vec2 - wav2vec 2.0 feature extractor is frozen.
speechbrain.lobes.models.huggingface_transformers.wav2vec2 - wav2vec 2.0 feature extractor is frozen.
speechbrain.core - Beginning experiment!
speechbrain.core - Experiment folder: results/zed_wavlm_large/78
speechbrain.dataio.encoder - Load called, but CategoricalEncoder is not empty. Loaded data will overwrite everything. This is normal if there is e.g. an unk label defined at init.
speechbrain.core - Gradscaler enabled: False. Using precision: fp32.
speechbrain.core - 311.3M trainable parameters in EmoDiaBrain
speechbrain.utils.checkpoints - Would load a checkpoint here, but none found yet.
speechbrain.utils.epoch_loop - Going into epoch 1
  0%|                                                                                                                                                                                                                                                     | 0/203 [00:00<?, ?it/s]Traceback (most recent call last):
  File "/cs/labs/oabend/avishai.elma/speechbrain/recipes/ZaionEmotionDataset/emotion_diarization/train.py", line 342, in <module>
    emo_id_brain.fit(
  File "/cs/labs/oabend/avishai.elma/speechbrain/speechbrain/core.py", line 1488, in fit
    self._fit_train(train_set=train_set, epoch=epoch, enable=enable)
  File "/cs/labs/oabend/avishai.elma/speechbrain/speechbrain/core.py", line 1316, in _fit_train
    loss = self.fit_batch(batch)
  File "/cs/labs/oabend/avishai.elma/speechbrain/speechbrain/core.py", line 1130, in fit_batch
    outputs = self.compute_forward(batch, sb.Stage.TRAIN)
  File "/cs/labs/oabend/avishai.elma/speechbrain/recipes/ZaionEmotionDataset/emotion_diarization/train.py", line 27, in compute_forward
    outputs = self.modules.wav2vec2(wavs)
  File "/cs/labs/oabend/avishai.elma/lab_env_v2/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/cs/labs/oabend/avishai.elma/lab_env_v2/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/cs/labs/oabend/avishai.elma/lab_env_v2/lib/python3.9/site-packages/torch/nn/parallel/distributed.py", line 1515, in forward
    inputs, kwargs = self._pre_forward(*inputs, **kwargs)
  File "/cs/labs/oabend/avishai.elma/lab_env_v2/lib/python3.9/site-packages/torch/nn/parallel/distributed.py", line 1409, in _pre_forward
    if torch.is_grad_enabled() and self.reducer._rebuild_buckets():
RuntimeError: Expected to have finished reduction in the prior iteration before starting a new one. This error indicates that your module has parameters that were not used in producing loss. You can enable unused parameter detection by passing the keyword argument `find_unused_parameters=True` to `torch.nn.parallel.DistributedDataParallel`, and by 
making sure all `forward` function outputs participate in calculating loss. 
If you already have done the above, then the distributed data parallel module wasn't able to locate the output tensors in the return value of your module's `forward` function. Please include the loss function and the structure of the return value of `forward` of your module when reporting this issue (e.g. list, dict, iterable).
Parameters which did not receive grad for rank 1: model.encoder.layers.23.final_layer_norm.bias, model.encoder.layers.23.final_layer_norm.weight, model.encoder.layers.23.feed_forward.output_dense.bias, model.encoder.layers.23.feed_forward.output_dense.weight, model.encoder.layers.23.feed_forward.intermediate_dense.bias, model.encoder.layers.23.feed_forward.intermediate_dense.weight, model.encoder.layers.23.layer_norm.bias, model.encoder.layers.23.layer_norm.weight, model.encoder.layers.23.attention.gru_rel_pos_linear.bias, model.encoder.layers.23.attention.gru_rel_pos_linear.weight, model.encoder.layers.23.attention.out_proj.bias, model.encoder.layers.23.attention.out_proj.weight, model.encoder.layers.23.attention.q_proj.bias, model.encoder.layers.23.attention.q_proj.weight, model.encoder.layers.23.attention.v_proj.bias, model.encoder.layers.23.attention.v_proj.weight, model.encoder.layers.23.attention.k_proj.bias, model.encoder.layers.23.attention.k_proj.weight, model.encoder.layers.23.attention.gru_rel_pos_const, model.encoder.layers.5.final_layer_norm.bias, model.encoder.layers.5.final_layer_norm.weight, model.encoder.layers.5.feed_forward.output_dense.bias, model.encoder.layers.5.feed_forward.output_dense.weight, model.encoder.layers.5.feed_forward.intermediate_dense.bias, model.encoder.layers.5.feed_forward.intermediate_dense.weight, model.encoder.layers.5.layer_norm.bias, model.encoder.layers.5.layer_norm.weight, model.encoder.layers.5.attention.gru_rel_pos_linear.bias, model.encoder.layers.5.attention.gru_rel_pos_linear.weight, model.encoder.layers.5.attention.out_proj.bias, model.encoder.layers.5.attention.out_proj.weight, model.encoder.layers.5.attention.q_proj.bias, model.encoder.layers.5.attention.q_proj.weight, model.encoder.layers.5.attention.v_proj.bias, model.encoder.layers.5.attention.v_proj.weight, model.encoder.layers.5.attention.k_proj.bias, model.encoder.layers.5.attention.k_proj.weight, model.encoder.layers.5.attention.gru_rel_pos_const, model.masked_spec_embed
Parameter indices which did not receive grad for rank 1: 0 106 107 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 448 449 450 451 452 453 454 455 456 457 458 459 460 461 462 463 464 465 466
  0%|█                                                                                                                                                                                                                           | 1/203 [00:02<07:19,  2.18s/it, train_loss=1.48]
speechbrain.core - Exception:
Traceback (most recent call last):
  File "/cs/labs/oabend/avishai.elma/speechbrain/recipes/ZaionEmotionDataset/emotion_diarization/train.py", line 342, in <module>
    emo_id_brain.fit(
  File "/cs/labs/oabend/avishai.elma/speechbrain/speechbrain/core.py", line 1488, in fit
    self._fit_train(train_set=train_set, epoch=epoch, enable=enable)
  File "/cs/labs/oabend/avishai.elma/speechbrain/speechbrain/core.py", line 1316, in _fit_train
    loss = self.fit_batch(batch)
  File "/cs/labs/oabend/avishai.elma/speechbrain/speechbrain/core.py", line 1130, in fit_batch
    outputs = self.compute_forward(batch, sb.Stage.TRAIN)
  File "/cs/labs/oabend/avishai.elma/speechbrain/recipes/ZaionEmotionDataset/emotion_diarization/train.py", line 27, in compute_forward
    outputs = self.modules.wav2vec2(wavs)
  File "/cs/labs/oabend/avishai.elma/lab_env_v2/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/cs/labs/oabend/avishai.elma/lab_env_v2/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/cs/labs/oabend/avishai.elma/lab_env_v2/lib/python3.9/site-packages/torch/nn/parallel/distributed.py", line 1515, in forward
    inputs, kwargs = self._pre_forward(*inputs, **kwargs)
  File "/cs/labs/oabend/avishai.elma/lab_env_v2/lib/python3.9/site-packages/torch/nn/parallel/distributed.py", line 1409, in _pre_forward
    if torch.is_grad_enabled() and self.reducer._rebuild_buckets():
RuntimeError: Expected to have finished reduction in the prior iteration before starting a new one. This error indicates that your module has parameters that were not used in producing loss. You can enable unused parameter detection by passing the keyword argument `find_unused_parameters=True` to `torch.nn.parallel.DistributedDataParallel`, and by 
making sure all `forward` function outputs participate in calculating loss. 
If you already have done the above, then the distributed data parallel module wasn't able to locate the output tensors in the return value of your module's `forward` function. Please include the loss function and the structure of the return value of `forward` of your module when reporting this issue (e.g. list, dict, iterable).
Parameters which did not receive grad for rank 0: model.encoder.layers.23.final_layer_norm.bias, model.encoder.layers.23.final_layer_norm.weight, model.encoder.layers.23.feed_forward.output_dense.bias, model.encoder.layers.23.feed_forward.output_dense.weight, model.encoder.layers.23.feed_forward.intermediate_dense.bias, model.encoder.layers.23.feed_forward.intermediate_dense.weight, model.encoder.layers.23.layer_norm.bias, model.encoder.layers.23.layer_norm.weight, model.encoder.layers.23.attention.gru_rel_pos_linear.bias, model.encoder.layers.23.attention.gru_rel_pos_linear.weight, model.encoder.layers.23.attention.out_proj.bias, model.encoder.layers.23.attention.out_proj.weight, model.encoder.layers.23.attention.q_proj.bias, model.encoder.layers.23.attention.q_proj.weight, model.encoder.layers.23.attention.v_proj.bias, model.encoder.layers.23.attention.v_proj.weight, model.encoder.layers.23.attention.k_proj.bias, model.encoder.layers.23.attention.k_proj.weight, model.encoder.layers.23.attention.gru_rel_pos_const, model.encoder.layers.5.final_layer_norm.bias, model.encoder.layers.5.final_layer_norm.weight, model.encoder.layers.5.feed_forward.output_dense.bias, model.encoder.layers.5.feed_forward.output_dense.weight, model.encoder.layers.5.feed_forward.intermediate_dense.bias, model.encoder.layers.5.feed_forward.intermediate_dense.weight, model.encoder.layers.5.layer_norm.bias, model.encoder.layers.5.layer_norm.weight, model.encoder.layers.5.attention.gru_rel_pos_linear.bias, model.encoder.layers.5.attention.gru_rel_pos_linear.weight, model.encoder.layers.5.attention.out_proj.bias, model.encoder.layers.5.attention.out_proj.weight, model.encoder.layers.5.attention.q_proj.bias, model.encoder.layers.5.attention.q_proj.weight, model.encoder.layers.5.attention.v_proj.bias, model.encoder.layers.5.attention.v_proj.weight, model.encoder.layers.5.attention.k_proj.bias, model.encoder.layers.5.attention.k_proj.weight, model.encoder.layers.5.attention.gru_rel_pos_const, model.masked_spec_embed
Parameter indices which did not receive grad for rank 0: 0 106 107 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 448 449 450 451 452 453 454 455 456 457 458 459 460 461 462 463 464 465 466
[2024-01-18 13:11:28,338] torch.distributed.elastic.multiprocessing.api: [ERROR] failed (exitcode: 1) local_rank: 0 (pid: 2492183) of binary: /cs/labs/oabend/avishai.elma/lab_env_v2/bin/python
Traceback (most recent call last):
  File "/cs/labs/oabend/avishai.elma/lab_env_v2/bin/torchrun", line 8, in <module>
    sys.exit(main())
  File "/cs/labs/oabend/avishai.elma/lab_env_v2/lib/python3.9/site-packages/torch/distributed/elastic/multiprocessing/errors/__init__.py", line 346, in wrapper
    return f(*args, **kwargs)
  File "/cs/labs/oabend/avishai.elma/lab_env_v2/lib/python3.9/site-packages/torch/distributed/run.py", line 806, in main
    run(args)
  File "/cs/labs/oabend/avishai.elma/lab_env_v2/lib/python3.9/site-packages/torch/distributed/run.py", line 797, in run
    elastic_launch(
  File "/cs/labs/oabend/avishai.elma/lab_env_v2/lib/python3.9/site-packages/torch/distributed/launcher/api.py", line 134, in __call__
    return launch_agent(self._config, self._entrypoint, list(args))
  File "/cs/labs/oabend/avishai.elma/lab_env_v2/lib/python3.9/site-packages/torch/distributed/launcher/api.py", line 264, in launch_agent
    raise ChildFailedError(
torch.distributed.elastic.multiprocessing.errors.ChildFailedError: 
============================================================
train.py FAILED
------------------------------------------------------------
Failures:
[1]:
  time      : 2024-01-18_13:11:28
  host      : arion-02.cs.huji.ac.il
  rank      : 1 (local_rank: 1)
  exitcode  : 1 (pid: 2492184)
  error_file: <N/A>
  traceback : To enable traceback see: https://pytorch.org/docs/stable/elastic/errors.html
------------------------------------------------------------
Root Cause (first observed failure):
[0]:
  time      : 2024-01-18_13:11:28
  host      : arion-02.cs.huji.ac.il
  rank      : 0 (local_rank: 0)
  exitcode  : 1 (pid: 2492183)
  error_file: <N/A>
  traceback : To enable traceback see: https://pytorch.org/docs/stable/elastic/errors.html
============================================================
```

**Comment by BenoitWang** [comment]
Hi @avishaiElmakies I was able to reproduce the issue. I did some research and this seems a HF issue for quite a while, not sure why we still have it... See [here](https://github.com/huggingface/transformers/issues/11452#issuecomment-851603295) and [here](https://discuss.huggingface.co/t/wav2vec-fine-tuning-with-multigpu/4894/14). 

So the problem is that wav2vec2-series models skip layers through its LayerDrop... By adding the following code in `speechbrain.core` after `scaled_loss.backward()`, you'll see that for each iteration some random layers are dropped, and when other gpus try to continue they get out of sync.
```
for name, param in self.modules.wav2vec2.named_parameters():
    if param.grad is None:
        print(name)
```

The solution is just adding `emo_id_brain.find_unused_parameters=True` before training. This may not be the neatest but the easiest for now.

**Comment by avishaiElmakies** [comment]
@BenoitWang thanks for looking into it.
 Is there maybe a better way to handle this? As far as I can see a simple config change or a boolean in a constructor might fix this (at least for the time being). 
The best solution will be to sync the processes, but this does sound hard and probably needs to be Done on HF side. 
I might be able to do it myself, but I would love if someone could look at it.

**Comment by Adel-Moumen** [comment]
This is weird since using --find_unused_parameters=True does the job when fine-tuning an SSL models (at least on ASR). 

Could you please retry with --find_unused_parameters=True enabled and let me know what your results are ?

**Comment by Adel-Moumen** [comment]
Hmm I don't think there's a better solution than that. As explained by @BenoitWang, w2v2 from HF is using a layer dropout strategy which causes issues with DDP (and that is why you really need to specify `--find_unused_parameters=True`). If you want to get some speedup, you can use `--precision=fp16` and/or increase the `--grad_accumulation_factor`. 

Do you know @TParcollet if there's ways to remove this layer drop ?

**Comment by TParcollet** [comment]
I've always used --find-unused-parameters. Tons of people want to use HuggingFace models, so they need to play with HF rules and the way they implement their code. This thing is out of control of SpeechBrain as it lies in how they implement their transformers library. @avishaiElmakies  The only solution is to deactivate layer_drop in the config.json of the HuggingFace model ;) You can also pass it to the HF class in the yaml I believe. But this may result (or may not) in degraded downstream performance..

**Comment by avishaiElmakies** [comment]
@TParcollet OK thanks!
